examples/src/main/java/com/mapbox/navigation/examples/core/WaypointsRerouteActivity.kt

```diff
@@ -0,0 +1,366 @@
+                .build()
+
+        mapboxNavigation?.navigationOptions?.locationEngine?.requestLocationUpdates(
+            requestLocationUpdateRequest,
+            locationListenerCallback,
+            mainLooper
+        )
+    }
+
+    private fun stopLocationUpdates() {
+        mapboxNavigation?.navigationOptions?.locationEngine?.removeLocationUpdates(
+            locationListenerCallback
+        )
+    }
+
+    private fun updateCameraOnNavigationStateChange(navigationStarted: Boolean) {
+        navigationMapboxMap?.apply {
+            if (navigationStarted) {
+                updateCameraTrackingMode(NavigationCamera.NAVIGATION_TRACKING_MODE_GPS)
+                updateLocationLayerRenderMode(RenderMode.GPS)
+            } else {
+                updateCameraTrackingMode(NavigationCamera.NAVIGATION_TRACKING_MODE_NONE)
+                updateLocationLayerRenderMode(RenderMode.COMPASS)
+            }
+        }
+    }
+
+    override fun onSaveInstanceState(outState: Bundle) {
+        super.onSaveInstanceState(outState)
+
+        // This is not the most efficient way to preserve the route on a device rotation.
+        // This is here to demonstrate that this event needs to be handled in order to
+        // redraw the route line after a rotation.
+        directionRoute?.let {
+            outState.putString(Utils.PRIMARY_ROUTE_BUNDLE_KEY, it.toJson())
+        }
+    }
+
+    override fun onRestoreInstanceState(savedInstanceState: Bundle?) {
+        super.onRestoreInstanceState(savedInstanceState)
+        directionRoute = Utils.getRouteFromBundle(savedInstanceState)
+    }
+
+    override fun onOffRouteStateChanged(offRoute: Boolean) {
+        if (offRoute) {
+            Toast.makeText(this, "You're off-route", Toast.LENGTH_SHORT).show()
+        }
+    }
+
+    override fun onRerouteStateChanged(rerouteState: RerouteState) {
+        val message: String = when (rerouteState) {
+            RerouteState.Idle -> "Reroute is idle"
+            RerouteState.Interrupted -> "Reroute request interrupted"
+            is RerouteState.Failed -> "Reroute request failed: ${rerouteState.message}"
+            RerouteState.FetchingRoute -> "Reroute request is in progress"
+            RerouteState.RouteFetched -> "Reroute request is finished successful"
+        }
+        Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
+    }
+
+    private class MyLocationEngineCallback(activity: WaypointsRerouteActivity) :
+        LocationEngineCallback<LocationEngineResult> {
+
+        private val activityRef = WeakReference(activity)
+
+        override fun onSuccess(result: LocationEngineResult?) {
+            result?.locations?.firstOrNull()?.let {
+                val activity = activityRef.get() ?: return@let
+                activity.navigationMapboxMap?.updateLocation(result.lastLocation)
+            }
+        }
+
+        override fun onFailure(exception: Exception) {
+        }
+    }
+
+    private class CoordinatesHolder {
```

Comment:
could you put classes `CoordinatesHolder` and `WaypointsController` on the same level? I mean it might be into our out of main class